### PR TITLE
Ian Cote - session01 basic homework assignment

### PR DESCRIPTION
**Pull request by iancote**
Howdy!

Here's the base homework assignment.  I'll look at doing another pull request with the "challenging" version if I get a chance to finish it (stuck traveling on Monday).

Thanks,
-ian

### CODE REVIEW

**Comment by iancote** [comment]
Hello!

I've added the bits for the select version.  In full disclosure, I ran across this awesome write up for dealing with exactly this scenario -  https://pymotw.com/2/select/  It was for python2x (or earlier?) so it needed updating, but very clearly explained everything.  My code is a pretty bald lift from that example.  It still passes tests.py as well as the load tests I ran.  Tracking in wireshark + the logging showed it handling multiple connections at a time.

Thanks,
-ian

**Comment by iancote** [comment]
Here's session02's homework.  I was able to do all the additional steps with the caveat that I altered the tests.py script so tests were accurate for the requirements for all but the last one (make_time.py), so you'll get two fails when running tests.py.  Everything else seems to be working ok.  Gotta love wireshark.

Thanks,
-ian

File: resources/session02/homework/http_server.py [hunk 1]
```diff
@@ -1,14 +1,32 @@
...
+
+    # Build content header
+    content_header = 'Content-Type: '.encode('utf8')
+    content_header += mimetype
+
+    # Build date header
+    gmt_date = 'Date: {}'.format(email.utils.formatdate()).encode('utf8')
+
+    # Build content length
+    content_length = 'Content-Length: {}'.format(len(body)).encode('utf8')
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: very good.  All correct.

File: resources/session02/homework/http_server.py [hunk 2]
```diff
@@ -17,25 +35,69 @@ def response_method_not_allowed():
     resp = []
     resp.append("HTTP/1.1 405 Method Not Allowed")
     resp.append("")
+    resp.append("")
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: This should not be required.  Is there a reason you are doing it?

File: resources/session02/homework/http_server.py [hunk 3]
```diff
@@ -17,25 +35,69 @@ def response_method_not_allowed():
...
     return "\r\n".join(resp).encode('utf8')
 
 
 def response_not_found():
     """returns a 404 Not Found response"""
-    return b""
+    resp = []
+    resp.append("HTTP/1.1 404 Not Found")
+    resp.append("")
+    resp.append("")
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: Again, why the extra blank line?  Only one should be required.

File: resources/session02/homework/http_server.py [hunk 4]
```diff
@@ -17,25 +35,69 @@ def response_method_not_allowed():
...
+    resp = []
+    resp.append("HTTP/1.1 500 Server Error")
+    resp.append("")
+    resp.append("")
+    return "\r\n".join(resp).encode('utf8')
 
 
 def parse_request(request):
     first_line = request.split("\r\n", 1)[0]
+    print('Request: "{}"'.format(first_line))
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: You might want to remove debugging helpers like this before you submit.

File: resources/session02/homework/http_server.py [hunk 5]
```diff
@@ -17,25 +35,69 @@ def response_method_not_allowed():
...
+def run_external_script(uri):
+    p = Popen([uri], stdout=PIPE)
+    return p.stdout.read()
+
+
 def resolve_uri(uri):
-    """This method should return appropriate content and a mime type"""
-    return b"still broken", b"text/plain"
+    """Return appropriate content and a mime type"""
+    uri = re.sub('\.\.', '', uri)  # pull any updirs out of the uri
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: This is a good security thought.  You might want to consider though if this is really what you want to do?  Relative uris with climbing to parent dirs are common in css and javascript paths and simply erasing them might be confusing.  

Given, of course, that this is a very simplistic HTTP server, it's probably not a big deal, and it's very good that you caught this possible weakness.  Just consider what other options you might have.

File: resources/session02/homework/http_server.py [hunk 6]
```diff
@@ -17,25 +35,69 @@ def response_method_not_allowed():
...
+        else:
+            with open(chroot_uri, 'rb') as f:
+                content = f.read()
+    elif os.path.isdir(chroot_uri):
+        # Send back dir listing
+        listing = os.listdir(chroot_uri)
+        content = '<!DOCTYPE html><title>Fancy Website</title><body>'
+        for item in listing:
+            # if looking for a uri in the root, kill the extra /
+            u = uri[1:] if uri == '/' else uri
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: This is interesting.  Did you find it necessary to do this?  Why?

File: resources/session02/homework/tests.py [hunk 7]
```diff
@@ -86,7 +86,8 @@ def test_passed_mimetype_in_response(self):
             ok = self.call_function_under_test(mimetype=expected)
             headers = extract_headers(ok)
             for header in headers:
-                name, value = header.split(b':')
+                # Added ", 1" to split to allow date header to pass
+                name, value = header.split(b':', 1)
```
Comment (cewing) [inline resources/session02/homework/tests.py]: heh.  good!

File: resources/session02/homework/tests.py [hunk 8]
```diff
@@ -208,7 +209,8 @@ def test_directory_resource(self):
         expected_names = [
             'a_web_page.html', 'images', 'make_time.py', 'sample.txt',
         ]
-        expected_mimetype = "text/plain"
+        # expected_mimetype = "text/plain" <- modified for html response
+        expected_mimetype = "text/html"
```
Comment (cewing) [inline resources/session02/homework/tests.py]: also good.

File: resources/session02/homework/tests.py [hunk 9]
```diff
@@ -306,7 +308,7 @@ def send_message(self, message, use_bytes=False):
         return response
 
     def test_get_request(self):
-        message = CRLF.join(['GET / HTTP/1.1', 'Host: example.com', ''])
+        message = CRLF.join(['GET / HTTP/1.1', 'Host: example.com', '', ''])
```
Comment (cewing) [inline resources/session02/homework/tests.py]: What is this extra line for?

File: resources/session02/homework/webroot/make_time.py [hunk 10]
```diff
@@ -19,7 +19,4 @@
 </http>
 """% time_str
 
-print html
-
-
-
+print(html)
```
Comment (cewing) [inline resources/session02/homework/webroot/make_time.py]: ahhhhhh.  Thanks.  My bad.

File: resources/session02/homework/http_server.py [hunk 11]
```diff
@@ -68,10 +130,13 @@ def server(log_buffer=sys.stderr):
...
                     except NameError:
                         response = response_not_found()
+                    except Exception as e:
+                        print('SERVER ERROR: {}'.format(e.message))
+                        response = server_error()
                     else:
                         response = response_ok(content, mime_type)
 
-                print('sending response', file=log_buffer)
+                print('sending response: {}'.format(response), file=log_buffer)
```
Comment (cewing) [inline resources/session02/homework/http_server.py]: This will result in **big** output to the terminal when responses include things like images.

**Comment by cewing** [comment]
@iancote, I've put comments inline on your PR here.  I'm going to close it as reviewed.  Thanks for your good work!

**Pull request closed without merging**